### Import section 

In [333]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as pl
import seaborn as sns
import scipy as sp


sns.set()
%matplotlib inline

from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVC, LinearSVC
from sklearn.ensemble import RandomForestClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.naive_bayes import GaussianNB
from sklearn.linear_model import Perceptron
from sklearn.linear_model import SGDClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.base import BaseEstimator


from sklearn.pipeline import Pipeline
from sklearn.feature_selection import VarianceThreshold
from sklearn.model_selection import train_test_split, GridSearchCV
from sklearn.neighbors import KNeighborsClassifier
from sklearn.preprocessing import Normalizer, LabelEncoder, StandardScaler, MinMaxScaler, MaxAbsScaler, OneHotEncoder
from sklearn.impute import SimpleImputer
from sklearn.compose import ColumnTransformer
from sklearn import set_config
from sklearn.base import TransformerMixin
from sklearn.preprocessing import OneHotEncoder



from sklearn.ensemble import RandomForestRegressor, GradientBoostingRegressor


In [334]:
storeFirstDataframe = pd.read_csv('./rawData/data_prueba_Forecasting.csv')

In [335]:
storeFirstDataframe.head(100)

Unnamed: 0 Formato Subformato  Tienda  Producto id_fec_diaria      Venta  \
0            0   Super     Basico       1    369772    2021-08-01   7.000000   
1            1   Super     Basico       1    369772    2021-08-02   6.000000   
2            2   Super     Basico       1    369772    2021-08-03   5.000000   
3            3   Super     Basico       1    369772    2021-08-04   4.000000   
4            4   Super     Basico       1    369772    2021-08-05   1.000000   
..         ...     ...        ...     ...       ...           ...        ...   
95          18   Hiper       Plus       1    303923    2021-08-19  54.124001   
96          19   Hiper       Plus       1    303923    2021-08-20  49.636001   
97          20   Hiper       Plus       1    303923    2021-08-21  82.614000   
98          21   Hiper       Plus       1    303923    2021-08-22  97.948002   
99          22   Hiper       Plus       1    303923    2021-08-23  80.686002   

        Precio   Venta_aa  Precio_aa  
0    17.168571   5.000000  21.063200  
1    17.141667   3.000000  21.388667  
2    17.136000   5.000000  21.376199  
3    16.995000   3.000000  21.360367  
4    17.240000   1.000000  21.397100  
..         ...        ...        ...  
95  112.441242  36.127999  94.703066  
96  112.436133  40.871000  95.028258  
97  112.669526  64.603002  94.451203  
98  112.506122  85.565998  94.756450  
99  112.906327  93.803000  94.868902  

[100 rows x 10 columns]

In [336]:
storeFirstDataframe.tail()

Unnamed: 0 Formato Subformato  Tienda  Producto id_fec_diaria  Venta  \
426372          87   Super       Plus       1   2098857    2021-10-27    8.0   
426373          88   Super       Plus       1   2098857    2021-10-28   18.0   
426374          89   Super       Plus       1   2098857    2021-10-29   13.0   
426375          90   Super       Plus       1   2098857    2021-10-30   17.0   
426376          91   Super       Plus       1   2098857    2021-10-31   23.0   

           Precio  Venta_aa  Precio_aa  
426372  27.003750      16.0  20.083056  
426373  26.911667      22.0  20.121114  
426374  24.463077      20.0  18.967170  
426375  25.404118      37.0  18.206573  
426376  23.815652      24.0  18.436345

In [337]:
storeFirstDataframe.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 426377 entries, 0 to 426376
Data columns (total 10 columns):
 #   Column         Non-Null Count   Dtype  
---  ------         --------------   -----  
 0   Unnamed: 0     426377 non-null  int64  
 1   Formato        426377 non-null  object 
 2   Subformato     426377 non-null  object 
 3   Tienda         426377 non-null  int64  
 4   Producto       426377 non-null  int64  
 5   id_fec_diaria  426377 non-null  object 
 6   Venta          426377 non-null  float64
 7   Precio         426377 non-null  float64
 8   Venta_aa       426377 non-null  float64
 9   Precio_aa      426377 non-null  float64
dtypes: float64(4), int64(3), object(3)
memory usage: 32.5+ MB



Como no tenemos valores nulos, no hay necesidad de hacer ningún tipo de data imputation.

Igualmente tenemos que el tipo de datos es justamente lo que requerimos para cada columna, excepto para las fechas que deberían ser tipo datetime, por ende tendremos qué cambiar las mismas.

Por ahora sólo cambiaré el nombre de la columna registrada como "unnamed: 0" por algo así como "día de venta".

In [338]:
storeFirstDataframe['id_fec_diaria'] = pd.to_datetime(storeFirstDataframe['id_fec_diaria'])
storeFirstDataframeDateChanged = storeFirstDataframe.copy()
storeFirstDataframeDateChanged.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 426377 entries, 0 to 426376
Data columns (total 10 columns):
 #   Column         Non-Null Count   Dtype         
---  ------         --------------   -----         
 0   Unnamed: 0     426377 non-null  int64         
 1   Formato        426377 non-null  object        
 2   Subformato     426377 non-null  object        
 3   Tienda         426377 non-null  int64         
 4   Producto       426377 non-null  int64         
 5   id_fec_diaria  426377 non-null  datetime64[ns]
 6   Venta          426377 non-null  float64       
 7   Precio         426377 non-null  float64       
 8   Venta_aa       426377 non-null  float64       
 9   Precio_aa      426377 non-null  float64       
dtypes: datetime64[ns](1), float64(4), int64(3), object(2)
memory usage: 32.5+ MB


In [339]:
storeFirstDataframeDateChanged.isnull().sum()

Unnamed: 0       0
Formato          0
Subformato       0
Tienda           0
Producto         0
id_fec_diaria    0
Venta            0
Precio           0
Venta_aa         0
Precio_aa        0
dtype: int64

In [340]:
storeFirstDataframeDateChanged.describe()

Unnamed: 0    Tienda      Producto          Venta         Precio  \
count  426377.000000  426377.0  4.263770e+05  426377.000000  426377.000000   
mean       45.552659       1.0  5.120868e+06      68.534217      61.782508   
std        26.532954       0.0  4.911189e+06     319.971154      70.248355   
min         0.000000       1.0  1.160000e+02       0.076000       0.870000   
25%        23.000000       1.0  1.035460e+06       4.000000      23.902857   
50%        46.000000       1.0  2.101039e+06      11.000000      40.270741   
75%        69.000000       1.0  1.120037e+07      38.000000      74.882001   
max        91.000000       1.0  1.153318e+07   41883.469971    1650.859985   

            Venta_aa      Precio_aa  
count  426377.000000  426377.000000  
mean       73.905263      58.161344  
std       348.099302      66.369279  
min         0.084000       0.008500  
25%         4.000000      22.561006  
50%        12.000000      38.382699  
75%        41.000000      68.401846  
max     46097.408234    1591.921753

In [341]:
storeFirstDataframeDateChanged.describe(include='all')

C:\Users\Alan Lopez\AppData\Local\Temp\ipykernel_350596\2635487392.py:1: FutureWarning: Treating datetime data as categorical rather than numeric in `.describe` is deprecated and will be removed in a future version of pandas. Specify `datetime_is_numeric=True` to silence this warning and adopt the future behavior now.
  storeFirstDataframeDateChanged.describe(include='all')


Unnamed: 0 Formato Subformato    Tienda      Producto  \
count   426377.000000  426377     426377  426377.0  4.263770e+05   
unique            NaN       2          3       NaN           NaN   
top               NaN   Hiper     Basico       NaN           NaN   
freq              NaN  323985     170484       NaN           NaN   
first             NaN     NaN        NaN       NaN           NaN   
last              NaN     NaN        NaN       NaN           NaN   
mean        45.552659     NaN        NaN       1.0  5.120868e+06   
std         26.532954     NaN        NaN       0.0  4.911189e+06   
min          0.000000     NaN        NaN       1.0  1.160000e+02   
25%         23.000000     NaN        NaN       1.0  1.035460e+06   
50%         46.000000     NaN        NaN       1.0  2.101039e+06   
75%         69.000000     NaN        NaN       1.0  1.120037e+07   
max         91.000000     NaN        NaN       1.0  1.153318e+07   

              id_fec_diaria          Venta         Precio       Venta_aa  \
count                426377  426377.000000  426377.000000  426377.000000   
unique                   92            NaN            NaN            NaN   
top     2021-08-14 00:00:00            NaN            NaN            NaN   
freq                   4761            NaN            NaN            NaN   
first   2021-08-01 00:00:00            NaN            NaN            NaN   
last    2021-10-31 00:00:00            NaN            NaN            NaN   
mean                    NaN      68.534217      61.782508      73.905263   
std                     NaN     319.971154      70.248355     348.099302   
min                     NaN       0.076000       0.870000       0.084000   
25%                     NaN       4.000000      23.902857       4.000000   
50%                     NaN      11.000000      40.270741      12.000000   
75%                     NaN      38.000000      74.882001      41.000000   
max                     NaN   41883.469971    1650.859985   46097.408234   

            Precio_aa  
count   426377.000000  
unique            NaN  
top               NaN  
freq              NaN  
first             NaN  
last              NaN  
mean        58.161344  
std         66.369279  
min          0.008500  
25%         22.561006  
50%         38.382699  
75%         68.401846  
max       1591.921753

In [342]:
ValoresSobreLaMediaDeVenta = storeFirstDataframeDateChanged[storeFirstDataframeDateChanged['Precio']>storeFirstDataframeDateChanged.Venta.mean()]['Producto'].unique().size
ValoresSobreLaMediaDeVenta

945

In [343]:
TotalProducts = storeFirstDataframeDateChanged.Producto.unique().shape[0]
TotalProducts

2405

In [344]:
uniqueProducts = pd.Series(storeFirstDataframe.Producto.unique())
uniqueProducts

0         369772
1         303923
2        1791075
3        1720010
4        1224404
          ...   
2400      378759
2401    11442635
2402    11400275
2403    11025261
2404     1155395
Length: 2405, dtype: int64

In [345]:
ValoresSobreLaMediaDeVenta/TotalProducts

0.39293139293139295

Tenemos que son 945 productos los que nos dan valores mayores a la media, que equivalen al 39%.  

Ahora vamos a llevar a cabo el one hot encoding para las categorías de formato y subformato.

In [346]:
(storeFirstDataframeDateChanged.duplicated() == 'true').count()


426377

In [347]:
duplicatedRows = storeFirstDataframeDateChanged.groupby(storeFirstDataframeDateChanged.columns.tolist(),as_index=False).size()
duplicatedRows


Unnamed: 0 Formato Subformato  Tienda  Producto id_fec_diaria  \
0                0   Hiper     Basico       1       116    2021-08-01   
1                0   Hiper     Basico       1       348    2021-08-01   
2                0   Hiper     Basico       1       407    2021-08-01   
3                0   Hiper     Basico       1       456    2021-08-01   
4                0   Hiper     Basico       1       480    2021-08-01   
...            ...     ...        ...     ...       ...           ...   
426372          91   Super       Plus       1  11498347    2021-10-31   
426373          91   Super       Plus       1  11499565    2021-10-31   
426374          91   Super       Plus       1  11501938    2021-10-31   
426375          91   Super       Plus       1  11515961    2021-10-31   
426376          91   Super       Plus       1  11526545    2021-10-31   

              Venta     Precio     Venta_aa  Precio_aa  size  
0       2335.494020  68.906064  2583.950001  48.858814     1  
1       2774.085018  27.362979  2609.015009  26.273482     1  
2       3653.395016  32.139943  4579.670032  26.264912     1  
3       6582.344982  18.987914  5421.385014  21.369818     1  
4       5486.099964  33.628015  5593.035027  28.991718     1  
...             ...        ...          ...        ...   ...  
426372     5.000000  27.512000     4.000000  27.424850     1  
426373     9.000000  67.826669     4.000000  57.490750     1  
426374     8.000000  23.821249     5.000000  19.696220     1  
426375     5.000000  66.176001     6.000000  75.181699     1  
426376   113.000000  30.424601   138.000000  33.702033     1  

[426377 rows x 11 columns]

In [349]:
total_rows = storeFirstDataframeDateChanged.shape[0]

In [353]:
storeFirstDataframeDateChanged['id_fec_diaria'] = storeFirstDataframeDateChanged['id_fec_diaria'].apply(lambda x: x.replace(year = x.year + 1))
storeSecondDataframeDateChanged

KeyError: "None of [DatetimeIndex(['2022-08-01', '2022-08-02', '2022-08-03', '2022-08-04',\n               '2022-08-05', '2022-08-06', '2022-08-09', '2022-08-10',\n               '2022-08-11', '2022-08-12',\n               ...\n               '2022-10-22', '2022-10-23', '2022-10-24', '2022-10-25',\n               '2022-10-26', '2022-10-27', '2022-10-28', '2022-10-29',\n               '2022-10-30', '2022-10-31'],\n              dtype='datetime64[ns]', length=426377, freq=None)] are in the [columns]"

Igualmente no tenemos ningún registro duplicado. Ahora lo que haremos es extraer los meses.

In [ ]:
list_months = []
list_days = []
list_years = []
list_months
list_days
list_years

[]

In [ ]:
for i in range(total_rows):
    list_months.append(storeFirstDataframeDateChanged['id_fec_diaria'][i].month)
    
for i in range(total_rows):
    list_days.append(storeFirstDataframeDateChanged['id_fec_diaria'][i].day)
    
for i in range(total_rows):
    list_years.append(storeFirstDataframeDateChanged['id_fec_diaria'][i].year)

In [ ]:
list_months = pd.Series(list_months)
list_days = pd.Series(list_days)
list_years = pd.Series(list_years)

In [ ]:
storeFirstDataframeDateChanged.head()

Unnamed: 0 Formato Subformato  Tienda  Producto id_fec_diaria  Venta  \
0           0   Super     Basico       1    369772    2021-08-01    7.0   
1           1   Super     Basico       1    369772    2021-08-02    6.0   
2           2   Super     Basico       1    369772    2021-08-03    5.0   
3           3   Super     Basico       1    369772    2021-08-04    4.0   
4           4   Super     Basico       1    369772    2021-08-05    1.0   

      Precio  Venta_aa  Precio_aa  
0  17.168571       5.0  21.063200  
1  17.141667       3.0  21.388667  
2  17.136000       5.0  21.376199  
3  16.995000       3.0  21.360367  
4  17.240000       1.0  21.397100

In [ ]:
storeFirstDataframeDateChanged['Mes'] = list_months
storeFirstDataframeDateChanged['Day'] = list_days
storeFirstDataframeDateChanged['Year'] = list_years

In [ ]:
storeFirstDataframeDateChanged.head()

Unnamed: 0 Formato Subformato  Tienda  Producto id_fec_diaria  Venta  \
0           0   Super     Basico       1    369772    2021-08-01    7.0   
1           1   Super     Basico       1    369772    2021-08-02    6.0   
2           2   Super     Basico       1    369772    2021-08-03    5.0   
3           3   Super     Basico       1    369772    2021-08-04    4.0   
4           4   Super     Basico       1    369772    2021-08-05    1.0   

      Precio  Venta_aa  Precio_aa  Mes  Day  Year  
0  17.168571       5.0  21.063200    8    1  2021  
1  17.141667       3.0  21.388667    8    2  2021  
2  17.136000       5.0  21.376199    8    3  2021  
3  16.995000       3.0  21.360367    8    4  2021  
4  17.240000       1.0  21.397100    8    5  2021

In [ ]:
storeFirstDataframeDateChanged = storeFirstDataframeDateChanged.drop('id_fec_diaria', axis=1)

In [ ]:
storeFirstDataframeFeatureDates = storeFirstDataframeDateChanged.copy()

In [ ]:
storeFirstDataframeFeatureDates.head()

Unnamed: 0 Formato Subformato  Tienda  Producto  Venta     Precio  \
0           0   Super     Basico       1    369772    7.0  17.168571   
1           1   Super     Basico       1    369772    6.0  17.141667   
2           2   Super     Basico       1    369772    5.0  17.136000   
3           3   Super     Basico       1    369772    4.0  16.995000   
4           4   Super     Basico       1    369772    1.0  17.240000   

   Venta_aa  Precio_aa  Mes  Day  Year  
0       5.0  21.063200    8    1  2021  
1       3.0  21.388667    8    2  2021  
2       5.0  21.376199    8    3  2021  
3       3.0  21.360367    8    4  2021  
4       1.0  21.397100    8    5  2021

In [ ]:
storeFirstDataframeFeatureDates = pd.get_dummies(storeFirstDataframeFeatureDates,
                                            columns = ['Formato', 'Subformato', 'Day', 'Mes', 'Year'])
storeFirstDataframeFeatureDates


Unnamed: 0  Tienda  Producto  Venta     Precio  Venta_aa  Precio_aa  \
0                0       1    369772    7.0  17.168571       5.0  21.063200   
1                1       1    369772    6.0  17.141667       3.0  21.388667   
2                2       1    369772    5.0  17.136000       5.0  21.376199   
3                3       1    369772    4.0  16.995000       3.0  21.360367   
4                4       1    369772    1.0  17.240000       1.0  21.397100   
...            ...     ...       ...    ...        ...       ...        ...   
426372          87       1   2098857    8.0  27.003750      16.0  20.083056   
426373          88       1   2098857   18.0  26.911667      22.0  20.121114   
426374          89       1   2098857   13.0  24.463077      20.0  18.967170   
426375          90       1   2098857   17.0  25.404118      37.0  18.206573   
426376          91       1   2098857   23.0  23.815652      24.0  18.436345   

        Formato_Hiper  Formato_Super  Subformato_Basico  ...  Day_26  Day_27  \
0                   0              1                  1  ...       0       0   
1                   0              1                  1  ...       0       0   
2                   0              1                  1  ...       0       0   
3                   0              1                  1  ...       0       0   
4                   0              1                  1  ...       0       0   
...               ...            ...                ...  ...     ...     ...   
426372              0              1                  0  ...       0       1   
426373              0              1                  0  ...       0       0   
426374              0              1                  0  ...       0       0   
426375              0              1                  0  ...       0       0   
426376              0              1                  0  ...       0       0   

        Day_28  Day_29  Day_30  Day_31  Mes_8  Mes_9  Mes_10  Year_2021  
0            0       0       0       0      1      0       0          1  
1            0       0       0       0      1      0       0          1  
2            0       0       0       0      1      0       0          1  
3            0       0       0       0      1      0       0          1  
4            0       0       0       0      1      0       0          1  
...        ...     ...     ...     ...    ...    ...     ...        ...  
426372       0       0       0       0      0      0       1          1  
426373       1       0       0       0      0      0       1          1  
426374       0       1       0       0      0      0       1          1  
426375       0       0       1       0      0      0       1          1  
426376       0       0       0       1      0      0       1          1  

[426377 rows x 47 columns]

In [ ]:
storeFirstDataframeFeatureDates.describe(include='all')

Unnamed: 0    Tienda      Producto          Venta         Precio  \
count  426377.000000  426377.0  4.263770e+05  426377.000000  426377.000000   
mean       45.552659       1.0  5.120868e+06      68.534217      61.782508   
std        26.532954       0.0  4.911189e+06     319.971154      70.248355   
min         0.000000       1.0  1.160000e+02       0.076000       0.870000   
25%        23.000000       1.0  1.035460e+06       4.000000      23.902857   
50%        46.000000       1.0  2.101039e+06      11.000000      40.270741   
75%        69.000000       1.0  1.120037e+07      38.000000      74.882001   
max        91.000000       1.0  1.153318e+07   41883.469971    1650.859985   

            Venta_aa      Precio_aa  Formato_Hiper  Formato_Super  \
count  426377.000000  426377.000000  426377.000000  426377.000000   
mean       73.905263      58.161344       0.759856       0.240144   
std       348.099302      66.369279       0.427171       0.427171   
min         0.084000       0.008500       0.000000       0.000000   
25%         4.000000      22.561006       1.000000       0.000000   
50%        12.000000      38.382699       1.000000       0.000000   
75%        41.000000      68.401846       1.000000       0.000000   
max     46097.408234    1591.921753       1.000000       1.000000   

       Subformato_Basico  ...         Day_26         Day_27         Day_28  \
count      426377.000000  ...  426377.000000  426377.000000  426377.000000   
mean            0.399843  ...       0.032434       0.032399       0.032886   
std             0.489867  ...       0.177149       0.177056       0.178339   
min             0.000000  ...       0.000000       0.000000       0.000000   
25%             0.000000  ...       0.000000       0.000000       0.000000   
50%             0.000000  ...       0.000000       0.000000       0.000000   
75%             1.000000  ...       0.000000       0.000000       0.000000   
max             1.000000  ...       1.000000       1.000000       1.000000   

              Day_29         Day_30         Day_31          Mes_8  \
count  426377.000000  426377.000000  426377.000000  426377.000000   
mean        0.033076       0.033083       0.022072       0.335914   
std         0.178836       0.178855       0.146918       0.472310   
min         0.000000       0.000000       0.000000       0.000000   
25%         0.000000       0.000000       0.000000       0.000000   
50%         0.000000       0.000000       0.000000       0.000000   
75%         0.000000       0.000000       0.000000       1.000000   
max         1.000000       1.000000       1.000000       1.000000   

               Mes_9         Mes_10  Year_2021  
count  426377.000000  426377.000000   426377.0  
mean        0.326955       0.337131        1.0  
std         0.469101       0.472731        0.0  
min         0.000000       0.000000        1.0  
25%         0.000000       0.000000        1.0  
50%         0.000000       0.000000        1.0  
75%         1.000000       1.000000        1.0  
max         1.000000       1.000000        1.0  

[8 rows x 47 columns]